test/test_transport.c

```diff
@@ -0,0 +1,155 @@
+
+#include "test_common.h"
+#include <sys/shm.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <librpmi.h>
+
+int g_num_xports;
+struct rpmi_transport *rpmi_transport_shmems[16];
+
+struct rpmi_shmem *rpmi_shmem_tests_create(const char *name, rpmi_uint64_t base,
+					  rpmi_uint32_t size);
+
+char *qemu_ram = NULL, *ctrl_base = NULL, *pub_mem_base = NULL;
+
+static inline void dump_msg(char *desc,  struct rpmi_message *rmsg)
+{
+	int i;
+	int *data = (int *)rmsg->data;
+
+	printf("%s: flags: %x, sgid: %x, svc_id: %x, dlen: %x, token: %x\n",
+		desc,
+		rmsg->header.flags,
+		rmsg->header.servicegroup_id,
+		rmsg->header.service_id,
+		rmsg->header.datalen, rmsg->header.token);
+
+	for (i = 0; i < (rmsg->header.datalen / 4); i++)
+		printf("data[%2d]: %08x \n", i, data[i]);
+}
+
+void *get_shm_addr(unsigned int xport_id, unsigned int *shm_sz)
+{
+	int shmid = 0;
+
+	if (!qemu_ram) {
+		shmid = shm_open("/qemu-ram", O_RDWR, 0);
+		if (shmid < 0) {
+			perror("shm_open(qemu-ram) failed");
+			return NULL;
+		}
+
+		qemu_ram = mmap(NULL,
+				SHM_SZ, PROT_READ | PROT_WRITE,
+				MAP_SHARED, shmid, 0);
+		if (qemu_ram == MAP_FAILED) {
+			perror("DDR Memory map failed");
+			return NULL;
+		}
+	}
+
+	*shm_sz = RPMI_SHM_SZ;
+
+	return qemu_ram + RPMI_SHM_OFSET + (RPMI_SHM_SZ * xport_id);
+}
+
+/* dump buffer in hexadecimal format */
+void hexdump(char *desc, unsigned int *buf, unsigned int len)
+{
+	unsigned int i;
+
+	printf("Dumping %s at %p\n", desc, buf);
+	for (i = 0; i < len/4; i++)
+		printf("%d: 0x%08x\n", i, *(buf+i));
+
+}
+
+int test_xport_dequeue(struct rpmi_transport *rpmi_xport, unsigned int qid, void *msg_data)
+{
+
+	int rc;
+
+	rc = rpmi_transport_dequeue(rpmi_xport, qid, msg_data);
+	if (!rc) {
+
+		//dump_msg((char *)"A2P messgege: ",  msg_data);
+		return 0;
+	}
+	return rc;
+}
+
+int test_xport_enqueue(struct rpmi_transport *rpmi_xport, unsigned int qid, void *req_msg)
+{
+	int rc;
+
+	//dump_msg((char *)"P2A message: ", req_msg);
+	do {
+		rc = rpmi_transport_enqueue(rpmi_xport, qid, req_msg);
+	} while (rc == RPMI_ERR_OUTOFRES);
+
+	if (rc) {
+		printf("%s: failed (error %d)\n",
+			__func__,  rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+struct rpmi_message *get_rpmi_msg(struct rpmi_transport *rpmi_xport)
+{
+	return calloc(1, (rpmi_xport->slot_size));
+}
+
+void free_rpmi_msg(struct rpmi_message *msg)
+{
+	if (msg)
+		free(msg);
+}
+
+struct rpmi_transport *init_test_rpmi_xport(int xport_id, int do_clear)
```

Comment:
This function is not need. Most of this things can be done in init_group().

In fact, you can move important stuff from test_transport.c and remove test_transport.c